### ISSUE TITLE
Macros

### ISSUE THREAD

**Issue by willrogers** [dls-controls/cs-web-proto#7]
We need to allow macros in PV names: `TS-TS-PV-01:${SUFFIX}`.

The macros dictionary will be kept in the store:

```
macros = {
  SUFFIX = "MYPV",
  ...
}
```

The middleware needs to be able to resolve the macros before making any connections. The component won't know that it's subscribed to the resolved PV name, so in order to get the right updates we need to be able to connect the unresolved macro name to the resolved macro name. This might be best done by keeping another dict in the store:

```
resolvedMacros = {
  TS-TS-PV-01:${SUFFIX}: "TS-TS-PV-01:MYPV"
}
```

The middleware will need to:

* resolve the macro 
* subscribe to the resolved macro
* dispatch the resolved macro so that the store can be updated later

**Comment by talwrii** [comment]
I don't quite understand what this is for; could I get a little more context?

**Comment by talwrii** [comment]
According to the [CA library specification](https://epics.anl.gov/base/R3-14/12-docs/CAref.html):

**int ca_clear_subscription ( evid EVID );

 Cancel a subscription.

All cancel-subscription requests such as the above are accumulated (buffered) and not forwarded to the server until one of ca_flush_io(), ca_pend_io(), ca_pend_event(), or ca_sg_block() are called. This allows several requests to be efficiently sent together in one message.

Note: This operation blocks until any user callbacks for this channel have run to completion. If callbacks take a lock (mutex) then it is the user's responsibility to ensure that this lock is not held when ca_clear_subscription() is called, otherwise a deadlock may ensue. (See also ca_clear_channel().)**

**Comment by willrogers** [comment]
Firstly let's check the race conditions.

React and Redux help a lot with these. There is one global store which is the source of truth. The widgets will display what is in the store.

Suppose the PV is `PREFIX:${MACRO}` and `MACRO` changes from `SUFFIX1` to `SUFFIX2`.

When a macro changes a new PV will be subscribed to. Updates will then come in for both PVs. Values will then be in the store for both PVs.

The component then needs to listen to the new PV rather than the old one. To do this the component needs to be listening to `PREFIX:${MACRO}` rather than `PREFIX:SUFFIX1`.

We update the store to the following:

````
store: {
  resolvedPVs: {
    PREFIX:${MACRO} = "PREFIX:SUFFIX2"
  },
  pvValues: {
    PREFIX:SUFFIX1 = value1,
    PREFIX:SUFFIX2 = value2
  }
}
```

So updating the `resolvedPVs` value will make the switch for us.

We may also like to unsubscribe from `PREFIX:SUFFIX1`.

**Comment by talwrii** [comment]
Hmm... I think I might like to unit test this one!

Yep, I agree we have a single store so we just need to update this prior to subscription. There's a bit of fiddling if we swap around resolutions (I think, I'm just going to unsubscribe and resubscribe each time I think). 

There are also nasty edge cases with duplicate subscriptions.

**Comment by willrogers** [comment]
Duplicate subscriptions: the problem I see here is that

 - we have to keep track of who has subscribed to what otherwise
 - we might unsubscribe to something that someone else still wants

**Comment by talwrii** [comment]
So, I'm half way through implementing this.

I want a little more context.

**What is the intended use case?**

Could you go through a couple of full stack intended uses  - this helps me understand performance tradeoffs.

** What should happen if two things resolve to the same pv? **

We can ban this. If we don't there are some edge cases.

Should we unsubscribe and resubscribe if we get a collision like this so as to push out new data to the new subscription - or does this potentially create downtime? If we don't then we need to push an artificial update to the new component - are we happy doing this?

It's straightforward enough to only unsubscribe if nothing resolves to a name and test this.

**Comment by willrogers** [comment]
Use cases:

* macros provided at screen load time. This allows the same screen to connect to different PVs (e.g. a bunch of PVs from the same type of magnet but with a different prefix to distinguish which magnet)
* macros changed at run time. This allows switching dynamically which magnet you're showing for a whole screen, but that's not really the use case. More likely you have a graph and you want to select which plot to show. If you can change the macro at run time the widget will automatically update to the new PV

If two things resolve to the same PV they probably want to listen to the same PV so that should be ok.

**Comment by willrogers** [comment]
Here is the same screen showing data for two different magnets because the prefix is set differently for all the PVs.

![mag1](https://user-images.githubusercontent.com/2440055/63584610-33b7e700-c595-11e9-87cb-0c0799317844.png)
![mag2](https://user-images.githubusercontent.com/2440055/63584638-403c3f80-c595-11e9-8e54-7b0c4a4193a5.png)

**Comment by talwrii** [comment]
> If two things resolve to the same PV they probably want to listen to the same PV so that should be ok.

Yep, there's a question about what we should do when a new component subscribes.

We can:

i. Wait for new updates to come along; or
ii. Create a new artificial update ourselves; or
iii. Unsubscribe and resubscribe to create an update.

**Comment by willrogers** [comment]
I hadn't thought of that. We need the update and we shouldn't resubscribe.

I think the way it works is that the component is listening to `PREFIX:{MACRO}` and `MACRO=SUFFIX1`.

When the component asks for a value from the store, it asks for the resolved PV for `PREFIX:${MACRO}` then asks for the value for that resolved PV `PREFIX:SUFFIX`. `PREFIX:SUFFIX` is already subscribed so we get some data.

Does that work?

**Comment by talwrii** [comment]
> Does that work?

I don't think so. 

If this is achieved via *remapping* rather than a subscription then there is nothing that causes the remapped component to go and fetch a new value. The new value have to be pushed to the components.  I guess the easiest way to do this is via another update.

I think components with normally be getting *pushed* data (via react) rather than polling it.

**Comment by willrogers** [comment]
It depends when an update to a component is triggered. Since we are updating something in the store I think we might be ok but we'd have to test.

**Comment by talwrii** [comment]
Sure. I guess updating the appropriate value in the store is the same as triggering an update event.

**Comment by talwrii** [comment]
Also some muttering about names. What names do we want to use to refer to these resolved and unresolved pvs? 

I'm using the term resolve and unresolved... and I guess it is going to be difficult to an unresolved pv as anything other than a pv even if we did give it a different name.

**Comment by willrogers** [comment]
https://stackoverflow.com/questions/40386128/how-does-a-redux-connected-component-know-when-to-re-render

**Comment by talwrii** [comment]
Okay. I have a first cut at this. I think this is pretty much what I'm going to submit (subject to feedback), but I probably want to edge case check this into oblivion as errors here are going to be painful to debug.  I don't know if I should set up some better logging - I don't know if there are good logging frameworks that we should be using. I think that remapping resolved pv's or creating of renewing connections feels like something that we should be logging. I guess for a prototype we can just use console.log and these are pretty easy to replace at a later date.

Anyway, on to some code because I want to talk about it!

[Here is a branch](https://github.com/talwrii/cs-web-proto/tree/talwrii--macros)

We have separate class solely responsible for mapping strings. https://github.com/talwrii/cs-web-proto/blob/talwrii--macros/src/redux/pvResolver.ts  I thought this might be a little "ravioli code" / "incestuous" to start with - but given the complexity here I'm pretty happy I did this. 

This interface to this class is that you tell it to do things, and it tells you about new things that you need to subscribe to, unsubscribe to, or refresh. That interface isn't *too* incestuous, though it is a little involved.

I wonder if I should be using a word other than incestuous? Tightly coupled maybe. 

What else. This is called from the middleware *object* (we've turned middleware into an object) and we pass through a single call "mapMacro". I imagine we might want the application to read the macros at some stage... which then becomes a bit nasty. I guess we might implement some sort of readonly wrapper so that the middleware still has complete control over mapping (because mapping can change subscriptions).

I think this interface looks quite good.  https://github.com/talwrii/cs-web-proto/blob/talwrii--macros/src/redux/pvResolver.ts

**Comment by willrogers** [comment]
I wrote down some of this discussion in a new wiki page:

https://github.com/dls-controls/cs-web-proto/wiki/Macros
